lib/zlib.js

```diff
@@ -225,230 +108,330 @@ function zlibBufferSync(engine, buffer) {
+        throw new RangeError('Invalid memLevel: ' + opts.memLevel);
+      }
+    }
 
-function isValidFlushFlag(flag) {
-  return flag === constants.Z_NO_FLUSH ||
-         flag === constants.Z_PARTIAL_FLUSH ||
-         flag === constants.Z_SYNC_FLUSH ||
-         flag === constants.Z_FULL_FLUSH ||
-         flag === constants.Z_FINISH ||
-         flag === constants.Z_BLOCK;
-}
+    if (opts.strategy && isInvalidStrategy(opts.strategy))
+      throw new TypeError('Invalid strategy: ' + opts.strategy);
 
-const strategies = [constants.Z_FILTERED,
-                    constants.Z_HUFFMAN_ONLY,
-                    constants.Z_RLE,
-                    constants.Z_FIXED,
-                    constants.Z_DEFAULT_STRATEGY];
+    if (opts.dictionary) {
+      if (!(opts.dictionary instanceof Buffer)) {
+        throw new TypeError(
+          'Invalid dictionary: it should be a Buffer instance');
+      }
+    }
 
-// the Zlib class they all inherit from
-// This thing manages the queue of requests, and returns
-// true or false if there is anything in the queue when
-// you call the .write() method.
+    this._handle = new binding.Zlib(mode);
+    this._handle.onerror = zlibOnError.bind(this);
+    this._hadError = false;
 
-function Zlib(opts, mode) {
-  this._opts = opts = opts || {};
-  this._chunkSize = opts.chunkSize || constants.Z_DEFAULT_CHUNK;
+    var level = constants.Z_DEFAULT_COMPRESSION;
+    if (typeof opts.level === 'number') level = opts.level;
 
-  Transform.call(this, opts);
+    var strategy = constants.Z_DEFAULT_STRATEGY;
+    if (typeof opts.strategy === 'number') strategy = opts.strategy;
 
-  if (opts.flush && !isValidFlushFlag(opts.flush)) {
-    throw new Error('Invalid flush flag: ' + opts.flush);
-  }
-  if (opts.finishFlush && !isValidFlushFlag(opts.finishFlush)) {
-    throw new Error('Invalid flush flag: ' + opts.finishFlush);
-  }
+    this._handle.init(opts.windowBits || constants.Z_DEFAULT_WINDOWBITS,
+                      level,
+                      opts.memLevel || constants.Z_DEFAULT_MEMLEVEL,
+                      strategy,
+                      opts.dictionary);
 
-  this._flushFlag = opts.flush || constants.Z_NO_FLUSH;
-  this._finishFlushFlag = typeof opts.finishFlush !== 'undefined' ?
-    opts.finishFlush : constants.Z_FINISH;
+    this._buffer = Buffer.allocUnsafe(this._chunkSize);
+    this._offset = 0;
+    this._level = level;
+    this._strategy = strategy;
 
-  if (opts.chunkSize) {
-    if (opts.chunkSize < constants.Z_MIN_CHUNK) {
-      throw new Error('Invalid chunk size: ' + opts.chunkSize);
-    }
-  }
+    this.once('end', this.close);
 
-  if (opts.windowBits) {
-    if (opts.windowBits < constants.Z_MIN_WINDOWBITS ||
-        opts.windowBits > constants.Z_MAX_WINDOWBITS) {
-      throw new Error('Invalid windowBits: ' + opts.windowBits);
-    }
+    Object.defineProperty(this, '_closed', {
+      get: () => !this._handle,
+      configurable: true,
+      enumerable: true
+    });
```

Comment:
This would make a wonderful `get _closed() { return !this._handle; }` shorthand if you’re turning it into a class anyway :)